### PR TITLE
Add function pointer restriction feature

### PR DESCRIPTION
**Pull request by danpoe**
Replaces #5168

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: doc/cprover-manual/restrict-function-pointer.md [hunk 1]
```diff
@@ -0,0 +1,173 @@
+[CPROVER Manual TOC](../../)
+
+## Restricting function pointers
+
+### Motivation
+
+CBMC comes with a way to resolve calls to function pointers to direct function
```
Comment (hannes-steffenhagen-diffblue) [inline doc/cprover-manual/restrict-function-pointer.md]: The documentation needs to be updated to reference `goto-instrument` instead of `cbmc` (also usage examples need to be updated to emphasise that this is a two-four step process source →binary→instrument→cbmc

File: src/goto-programs/label_function_pointer_call_sites.cpp [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+          function_call_site_symbol.mode = goto_function_symbol_mode;
+          return function_call_site_symbol;
+        }());
+        auto const new_function_pointer =
+          goto_model.symbol_table.lookup_ref(new_symbol_name).symbol_expr();
+        auto const assign_instruction = goto_programt::make_assignment(
+          code_assignt{new_function_pointer,
+                       function_pointer_dereference.pointer()},
+          source_location);
+        goto_function.second.body.insert_before(it, assign_instruction);
```
Comment (thk123) [inline src/goto-programs/label_function_pointer_call_sites.cpp:59]: :question: Is it safe to modify the collection we're iterating over?

File: src/goto-programs/label_function_pointer_call_sites.cpp [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        to_code_function_call(it->code).function() =
+          dereference_exprt{new_function_pointer};
+      });
```
Comment (thk123) [inline src/goto-programs/label_function_pointer_call_sites.cpp]: :bulb: I find this kind of code difficult to work out what it is doing as it involves a lot of straight line code. 

```C++
const auto is_function_pointer_call = [](const goto_programt::targett it) {
        return it->is_function_call() && can_cast_expr<dereference_exprt>(
                                           it->get_function_call().function());
      };
struct required_labelt
{
  symbolt new_symbol;
  goto_programt::targett function_pointer_call;
}
  

std::vector<required_labelt> fp_labels = make_range(goto_function.body.instruction)
  .filter(is_function_pointer_call)
  .map([&](goto_programt::targett function_pointer_call) {
    .. 
   const label_symbol = [&] {
          symbolt function_call_site_symbol{};
          function_call_site_symbol.name = function_call_site_symbol.base_name =
            function_call_site_symbol.pretty_name = new_symbol_name;
          function_call_site_symbol.type =
            function_pointer_dereference.pointer().type();
          function_call_site_symbol.location = function_call.source_location();
          function_call_site_symbol.is_lvalue = true;
          function_call_site_symbol.mode = goto_function_symbol_mode;
          return function_call_site_symbol;
        }();
   return required_labelt{label_symbol, function_pointer_call};).collect();

for(const auto new_label : required_labels)
{
  goto_model.symbol_table.insert(new_label.label_symbol);
  goto_function.second.body.insert_before(new_label.function_pointer_call, new_label.assign_instruction);
}
```

or something like that? It breaks up the extracting the data from the instruction, and the mutations that we need to perform. It also means your not modifying a collection your iterating over, which always makes me nervous

File: src/goto-programs/label_function_pointer_call_sites.cpp [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                                           it->get_function_call().function());
+      },
+      [&](goto_programt::targett it) {
+        auto const &function_call = it->get_function_call();
+        auto const &function_pointer_dereference =
+          to_dereference_expr(function_call.function());
+        auto const &source_location = function_call.source_location();
+        auto const &goto_function_symbol_mode =
+          goto_model.symbol_table.lookup_ref(goto_function.first).mode;
+        auto const new_symbol_name =
```
Comment (thk123) [inline src/goto-programs/label_function_pointer_call_sites.cpp]: `new_symbol_name` could be better - what is the symbol for? Is it the label? `call_site_symbol_name`?

File: src/goto-programs/label_function_pointer_call_sites.cpp [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+\*******************************************************************/
+
+/// \file
+/// Label function pointer call sites across a goto model
+
+#include "label_function_pointer_call_sites.h"
+
+#include <util/fresh_symbol.h>
+
+void label_function_pointer_call_sites(goto_modelt &goto_model)
```
Comment (thk123) [inline src/goto-programs/label_function_pointer_call_sites.cpp:13]: :no_entry_sign: A comment here demonstrating exactly the transformation might be helpful

```
///  Transform goto code of function pointer calls of the form:
/// (*fp)(...)
/// Into:
/// function_call_site_symbol = fp;
/// (*function_call_site_symbol)(...)
```

File: src/util/irep.h [hunk 6]
```diff
@@ -50,6 +50,11 @@ inline const std::string &id2string(const irep_idt &d)
   #endif
 }
 
+inline irep_idt string2id(const std::string &id_string)
```
Comment (thk123) [inline src/util/irep.h]: :question: I believe this is already an existing implicit conversion

File: src/goto-programs/label_function_pointer_call_sites.h [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+/// 2. A dereference of a symbol expression of a function pointer type
+///
+/// This makes following stages dealing with function calls easier, because they
+/// only need to be able to handle these two cases.
+///
+/// It does this by replacing all CALL instructions to function pointers with an
+/// assignment to a function pointer variable with a name following the pattern
+/// [function_name].function_pointer_call.[N], where "N" is the nth call to a
+/// function pointer in the function "function_name".
+void label_function_pointer_call_sites(goto_modelt &goto_model);
```
Comment (thk123) [inline src/goto-programs/label_function_pointer_call_sites.h:28]: :pick: This looks quite stand alone - it would be easier to review if this were in a separate commit. 
:bulb: As it is so isolated, I think a unit test with a few different kinds calls that get transformed would both make the code more robust and somewhat document the transformation

File: src/goto-programs/restrict_function_pointers.h [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+#define OPT_RESTRICT_FUNCTION_POINTER                                          \
+  "(" RESTRICT_FUNCTION_POINTER_OPT                                            \
+  "):"                                                                         \
+  "(" RESTRICT_FUNCTION_POINTER_FROM_FILE_OPT "):"
+
+#define RESTRICT_FUNCTION_POINTER_HELP                                         \
+  "--" RESTRICT_FUNCTION_POINTER_OPT                                           \
+  " <pointer_name>/<target[,targets]*>\n"                                      \
+  "           restrict a function pointer to a set of possible targets\n"      \
```
Comment (thk123) [inline src/goto-programs/restrict_function_pointers.h:41]: This would definitely benefit with some examples in some documentation somewhere

File: src/goto-programs/restrict_function_pointers.cpp [hunk 9]
```diff
@@ -0,0 +1,415 @@
...
+          }
+          return irep_idt{to_json_string(array_element).value};
+        });
+      return possible_targets;
+    }());
+  }
+  return function_pointer_restrictionst{restrictions};
+}
+
+void function_pointer_restrictionst::write_to_file(
```
Comment (thk123) [inline src/goto-programs/restrict_function_pointers.cpp]: :bulb: if this returned a json object then it would be possible to unit test this

File: src/goto-programs/restrict_function_pointers.cpp [hunk 10]
```diff
@@ -0,0 +1,415 @@
...
+    message_handler);
+  return {merge_function_pointer_restrictions(
+    file_restrictions, commandline_restrictions)};
+}
+
+function_pointer_restrictionst function_pointer_restrictionst::read_from_file(
+  const std::string &filename,
+  message_handlert &message_handler)
+{
+  auto failed = [](bool failFlag) { return failFlag; };
```
Comment (thk123) [inline src/goto-programs/restrict_function_pointers.cpp]: ~:question:~ :pick:  This name seems a little misleading - is this not just the identity function?

Comment (thk123) [reply src/goto-programs/restrict_function_pointers.cpp]: Ah I see what is being done with this... A comment might not be an awful idea, but not blocking

File: src/goto-programs/restrict_function_pointers.cpp [hunk 11]
```diff
@@ -0,0 +1,415 @@
...
+  if(failed(parse_json(inFile, filename, message_handler, json)))
+  {
+    throw system_exceptiont{
+      "failed to read function pointer restrictions from " + filename};
+  }
+  if(!json.is_object())
+  {
+    throw system_exceptiont{"top level item is not an object"};
+  }
+  for(auto const &kv_pair : to_json_object(json))
```
Comment (thk123) [inline src/goto-programs/restrict_function_pointers.cpp]: :bulb: Better naming: `read_restriction`?

Comment (thk123) [inline src/goto-programs/restrict_function_pointers.cpp]: :bulb: Ditto - if this read a json_objectt it would be easier to unit test

File: src/goto-programs/restrict_function_pointers.cpp [hunk 12]
```diff
@@ -0,0 +1,415 @@
...
+    auto const restrictions =
+      function_pointer_restrictionst::read_from_file(filename, message_handler);
+    merged_restrictions = merge_function_pointer_restrictions(
+      merged_restrictions, restrictions.restrictions);
+  }
+  return merged_restrictions;
+}
+} // namespace
+
+function_pointer_restrictionst function_pointer_restrictionst::from_options(
```
Comment (thk123) [inline src/goto-programs/restrict_function_pointers.cpp:357]: :bulb: A great candidate for unit tests (which would document expected format for free)

File: src/goto-programs/restrict_function_pointers.cpp [hunk 13]
```diff
@@ -0,0 +1,415 @@
...
+#include <ansi-c/expr2c.h>
+#include <json/json_parser.h>
+#include <util/expr_iterator.h>
+#include <util/string_utils.h>
+
+#include <algorithm>
+#include <fstream>
+#include <iostream>
+
+namespace
```
Comment (thk123) [inline src/goto-programs/restrict_function_pointers.cpp:21]: :question: What is the purpose of this `namespace`?

File: src/goto-programs/restrict_function_pointers.cpp [hunk 14]
```diff
@@ -0,0 +1,415 @@
...
+parse_function_pointer_restrictions_from_command_line(
+  const std::list<std::string> &restriction_opts)
+{
+  auto function_pointer_restrictions =
+    function_pointer_restrictionst::restrictionst{};
+  for(auto const &restriction_opt : restriction_opts)
+  {
+    // the format for restrictions is <pointer_name>/<target[,more_targets]*>
+    // exactly one pointer and at least one target
+    auto const pointer_name_end = restriction_opt.find('/');
```
Comment (thk123) [inline src/goto-programs/restrict_function_pointers.cpp]: :question: Could regex be used here? Rather than manually parsing the string

File: src/goto-programs/restrict_function_pointers.cpp [hunk 15]
```diff
@@ -0,0 +1,415 @@
...
+      }
+    }
+    if(!function_pointer_restrictions
+          .emplace(irep_idt{pointer_name}, target_names)
+          .second)
+    {
+      throw invalid_command_line_argument_exceptiont{
+        "function pointer restriction for `" + pointer_name +
+          "' was specified twice",
+        "--" RESTRICT_FUNCTION_POINTER_OPT};
```
Comment (thk123) [inline src/goto-programs/restrict_function_pointers.cpp:268]: :no_entry_sign: Pull out into a function that takes a string and returns a data structure containing pointer_name, target_names and unit test to cover all the awkward cases here (before swapping it to use regex if you do choose to do that)

File: src/goto-programs/restrict_function_pointers.cpp [hunk 16]
```diff
@@ -0,0 +1,415 @@
...
+    cmdline.get_values(RESTRICT_FUNCTION_POINTER_OPT));
+  options.set_option(
+    RESTRICT_FUNCTION_POINTER_FROM_FILE_OPT,
+    cmdline.get_values(RESTRICT_FUNCTION_POINTER_FROM_FILE_OPT));
+}
+
+namespace
+{
+function_pointer_restrictionst::restrictionst
+merge_function_pointer_restrictions(
```
Comment (thk123) [inline src/goto-programs/restrict_function_pointers.cpp]: might be worth a quick unit test for this

File: src/goto-programs/restrict_function_pointers.cpp [hunk 17]
```diff
@@ -0,0 +1,415 @@
...
+  const function_pointer_restrictionst::restrictionst &lhs,
+  const function_pointer_restrictionst::restrictionst &rhs)
+{
+  auto result = lhs;
```
Comment (thk123) [inline src/goto-programs/restrict_function_pointers.cpp]: :pick: Since you're immediately copying it, you could change `lhs` to be taken by value. That way if the user is doing

```
x = merge(x, y)
```
Then can do
```
x = merge(std::move(x), y)
```
and save a copy. 

It is also one less line of code!

File: src/goto-programs/restrict_function_pointers.cpp [hunk 18]
```diff
@@ -0,0 +1,415 @@
...
+
+void restrict_function_pointers(
+  goto_modelt &goto_model,
+  const function_pointer_restrictionst &restrictions)
+{
+  typecheck_function_pointer_restrictions(goto_model, restrictions);
+
+  for(auto &goto_function : goto_model.goto_functions.function_map)
+  {
+    auto &goto_function_body = goto_function.second.body;
```
Comment (thk123) [inline src/goto-programs/restrict_function_pointers.cpp]: This gets pretty indented - consider breakin gout the body of the loop into a function that deals with bodies to save one level of indentiation

File: src/goto-programs/restrict_function_pointers.cpp [hunk 19]
```diff
@@ -0,0 +1,415 @@
...
+              auto const replaced_instruction_location =
+                goto_function.second.body.insert_before(
+                  goto_end_if_location, new_instruction);
+              else_location = goto_function.second.body.insert_before(
+                replaced_instruction_location,
+                goto_programt::make_goto(
+                  else_location,
+                  notequal_exprt{
+                    pointer_symbol,
+                    address_of_exprt{function_pointer_target_symbol_expr}}));
```
Comment (thk123) [inline src/goto-programs/restrict_function_pointers.cpp]: Hard to read, suggestions to improve:

 - pull out bodies of loops to reduce indentation
 - use make_range / filter rather than lots of nested ifs
 - invert the ifs with a `continue` to reduce indentation
 - use smaller functions rather than comments to explain blocks of code

File: src/goto-programs/restrict_function_pointers.cpp [hunk 20]
```diff
@@ -0,0 +1,415 @@
...
+    // restrictions for, and if so branch on its value and insert concrete calls
+    // to the restriction functions
+    for_each_function_call(
+      goto_function.second, [&](const goto_programt::targett location) {
+        // Check if this is calling a function pointer, and if so if it is one
+        // we have a restriction for
+        const auto &original_function_call = location->get_function_call();
+        if(can_cast_expr<dereference_exprt>(original_function_call.function()))
+        {
+          // because we simplify before this stage a dereference can only
```
Comment (thk123) [inline src/goto-programs/restrict_function_pointers.cpp]: Probably worth an INVARIANT here incase someone removes the turning into canoncical form step (and ideally the comment would reference the simplification you are using. 

I also wouldn't call it `simplify` since that is a very specific thing.Instead ideally name that transformation `normalize_function_pointer_calls`? (Since we normalize function pointer calls)

Comment (thk123) [reply src/goto-programs/restrict_function_pointers.cpp]: Or using your existing words `since we already label function pointer call sites`

File: src/goto-programs/restrict_function_pointers.cpp [hunk 21]
```diff
@@ -0,0 +1,415 @@
...
+          auto const &called_function_pointer =
+            to_dereference_expr(original_function_call.function()).pointer();
+          PRECONDITION(can_cast_expr<symbol_exprt>(called_function_pointer));
+          auto const &pointer_symbol = to_symbol_expr(called_function_pointer);
+          auto const restriction_iterator =
+            restrictions.restrictions.find(pointer_symbol.get_identifier());
+          if(restriction_iterator != restrictions.restrictions.end())
+          {
+            auto const &restriction = *restriction_iterator;
+            // if we can, we will replace uses of it by the
```
Comment (thk123) [inline src/goto-programs/restrict_function_pointers.cpp]: :question: This comment seems to stop mid sentence?

File: src/goto-programs/restrict_function_pointers.cpp [hunk 22]
```diff
@@ -0,0 +1,415 @@
...
+            *location = goto_programt::make_assertion(
+              false_exprt{},
+              make_function_pointer_restriction_assertion_source_location(
+                original_function_call_instruction.source_location,
+                restriction));
+            auto const assume_false_location = goto_function_body.insert_after(
+              location,
+              goto_programt::make_assumption(
+                false_exprt{},
+                original_function_call_instruction.source_location));
```
Comment (thk123) [inline src/goto-programs/restrict_function_pointers.cpp]: :question: A why comment here would be helpful

File: src/goto-programs/restrict_function_pointers.cpp [hunk 23]
```diff
@@ -0,0 +1,415 @@
...
+            // this is mutable because we'll update this at the end of each
+            // loop iteration to always point at the start of the branch
+            // we created
+            auto else_location = location;
+            auto const end_if_location = goto_function_body.insert_after(
+              assume_false_location, goto_programt::make_skip());
+            for(auto const &restriction_target : restriction.second)
+            {
+              auto new_instruction = original_function_call_instruction;
```
Comment (thk123) [inline src/goto-programs/restrict_function_pointers.cpp]: :bulb: This definitively should be pulled out into a separate function, independently unit tested to get an idea of what goto code it produces.

File: src/goto-programs/restrict_function_pointers.cpp [hunk 24]
```diff
@@ -0,0 +1,415 @@
...
+    }
+    comment << ']';
+  }
+  source_location.set_comment(string2id(comment.str()));
+  source_location.set_property_class(ID_assertion);
+  return source_location;
+}
+
+template <typename Handler>
+void for_each_function_call(goto_functiont &goto_function, Handler handler)
```
Comment (thk123) [inline src/goto-programs/restrict_function_pointers.cpp:108]: :pick: I'd much prefer using the range framework rather than introducing these helper functions

File: src/goto-programs/restrict_function_pointers.cpp [hunk 25]
```diff
@@ -0,0 +1,415 @@
...
+      if(!first)
+      {
+        comment << ", ";
+      }
+      else
+      {
+        first = false;
+      }
+      comment << target;
+    }
```
Comment (thk123) [inline src/goto-programs/restrict_function_pointers.cpp]: :idea: `join_strings` handles this for you
```suggestion
    join_strings(comment, restrictions.second.begin(), restrictions.second.end(), ',');
```

File: src/goto-programs/restrict_function_pointers.cpp [hunk 26]
```diff
@@ -0,0 +1,415 @@
...
+
+#include <algorithm>
+#include <fstream>
+#include <iostream>
+
+namespace
+{
+void typecheck_function_pointer_restrictions(
+  const goto_modelt &goto_model,
+  const function_pointer_restrictionst &restrictions)
```
Comment (thk123) [inline src/goto-programs/restrict_function_pointers.cpp:25]: :green_heart: awesome checks - this is exactly the kind of things that would be super frustrating from user perspective to only find out how letting the analysis run!
Be great to have a unit test for each of the error checks here

File: src/goto-programs/restrict_function_pointers.cpp [hunk 27]
```diff
@@ -0,0 +1,415 @@
+/*******************************************************************\
```
Comment (thk123) [inline src/goto-programs/restrict_function_pointers.cpp:1]: This file is pretty big - I'd be tempted to split up the command line option parsing / error checking from the goto code transformation stuff

**Review by thk123 — COMMENTED** [review]
On the whole - I think this is a really good PR, lots of clear names, functions do what they say. It is easy to see how it all plugs together. Also a big fan of what seems to be some very robust user error handling :+1: 

Got a bunch of comments that basically come in two flavours:
 - lots of opportunities for unit tests which would give much greater confidence on the edge cases in the code and saves 10s of regression tests! 
 - some large blocks of code that are quite intimidating to start reading that I think a few simple refactors could greatly improve readability

I haven't yet reviewed the regression tests or documentation, but thought I'd leave the code comments so you could begin incorporating the ones you agree with

File: src/util/irep.h (hunk 6 above)
Comment (thk123) [reply src/util/irep.h]: This gets removed in later commit

File: regression/goto-instrument/restrict-function-pointer-to-complex-expression/test.desc [hunk 28]
```diff
@@ -0,0 +1,7 @@
+CORE
+test.c
+--restrict-function-pointer "use_fg.function_pointer_call.1/f,g"
+\[use_fg.assertion.2\] line \d+ assertion \(choice \? fptr : gptr\)\(10\) == 10 \+ choice: SUCCESS
+\[use_fg.assertion.1\] line \d+ dereferenced function pointer at use_fg.function_pointer_call.1 must be one of \[(f|g), (f|g)\]: SUCCESS
+^EXIT=0$
+^SIGNAL=0$
```
Comment (thk123) [inline regression/goto-instrument/restrict-function-pointer-to-complex-expression/test.desc:7]: :no_entry_sign: Please add explanations for each test into the desc file

File: doc/cprover-manual/restrict-function-pointer.md [hunk 29]
```diff
@@ -0,0 +1,180 @@
...
+through function pointers by case distinctions over calls to given sets of
+functions.
+
+### Motivation
+
+The CPROVER framework includes a goto program transformation pass
+`remove_function_pointers()` to resolve calls to function pointers to direct
+function calls. The pass is needed by `cbmc`, as symbolic execution itself can't
+handle calls to function pointers. In practice, the transformation pass works as
+follows:
```
Comment (thk123) [inline doc/cprover-manual/restrict-function-pointer.md:15]: :bulb: this is kind of an impl detail - and as such I'd be tempted to bury it at the _bottom_ of the docs

File: regression/goto-instrument/restrict-function-pointer-to-complex-expression/test.desc [hunk 30]
```diff
@@ -0,0 +1,7 @@
+CORE
+test.c
+--restrict-function-pointer "use_fg.function_pointer_call.1/f,g"
```
Comment (thk123) [inline regression/goto-instrument/restrict-function-pointer-to-complex-expression/test.desc:3]: :question: What is the point of `function_pointer_call`, why isn't the syntax just `use_fg.1/f,g`?
:question: Can I specify more all calls? e.g. `use_fg/f,g` for all calls inside `use_fg`?

File: regression/goto-instrument/restrict-function-pointer-to-complex-expression/test.c [hunk 31]
```diff
@@ -0,0 +1,59 @@
...
+{
+  assert((choice ? fptr : gptr)(10) == 10 + choice);
+}
+
+// this is just here to distinguish the behavior from FP removal, which'd include g
+int g_always_false_cond = 0;
+
+int main(void)
+{
+  use_fg(0, get_f(), get_g());
```
Comment (thk123) [inline regression/goto-instrument/restrict-function-pointer-to-complex-expression/test.c:15]: I'm a bit wary that constant propogation will make this example work for free? Consider using a nondet int for the `choice` param

File: regression/goto-instrument/restrict-function-pointer-to-complex-expression/test.c [hunk 32]
```diff
@@ -0,0 +1,59 @@
...
+
+fptr_t get_f(void);
+fptr_t get_g(void);
+
+void use_fg(int choice, fptr_t fptr, fptr_t gptr)
+{
+  assert((choice ? fptr : gptr)(10) == 10 + choice);
+}
+
+// this is just here to distinguish the behavior from FP removal, which'd include g
```
Comment (thk123) [inline regression/goto-instrument/restrict-function-pointer-to-complex-expression/test.c]: :question: I don't understand this comment, or this variable

File: regression/goto-instrument/restrict-function-pointer-to-multiple-functions-code-check/test.desc [hunk 33]
```diff
@@ -0,0 +1,9 @@
+CORE
+test.c
+--restrict-function-pointer use_f.function_pointer_call.1/f,g --show-goto-functions
+f\(10\)
```
Comment (thk123) [inline regression/goto-instrument/restrict-function-pointer-to-multiple-functions-code-check/test.desc:4]: :green_heart: this test seems a precise approach

File: regression/goto-instrument/restrict-function-pointer-to-multiple-functions-via-file-and-command-line-options/test.desc [hunk 34]
```diff
@@ -0,0 +1,6 @@
+CORE
+test.c
+--function-pointer-restrictions-file restrictions.json --restrict-function-pointer use_f.function_pointer_call.1/g
```
Comment (thk123) [inline regression/goto-instrument/restrict-function-pointer-to-multiple-functions-via-file-and-command-line-options/test.desc:3]: :question: Wouldn't this test work if it was ignoring the command line arguments too?

**Review by thk123 — COMMENTED** [review]
It is hard to see exactly what the tests are testing, I think a short sentence at the foot of each desc describing it's purpose, so appologies if there are tests for the following:

 - writing to a json file (I saw the code - not sure how it is to be used in this PR)
 - multiple distinct function pointer calls replaced
 - a function pointer call that _isn't_ replaced
 - multiple calls to the same function pointer being replaced by different fps
 - parsing a mismatched type (though think we should test all of the error checks fully through unit tests - this would just be to check the validation code is indeed being used)

File: regression/goto-instrument/restrict-function-pointer-to-complex-expression/test.desc (hunk 30 above)
Comment (hannes-steffenhagen-diffblue) [reply regression/goto-instrument/restrict-function-pointer-to-complex-expression/test.desc:3]: 1. This could be shortened, yes.
2. No, but something similar is in a follow up PR

File: regression/goto-instrument/restrict-function-pointer-to-complex-expression/test.c (hunk 32 above)
Comment (hannes-steffenhagen-diffblue) [reply regression/goto-instrument/restrict-function-pointer-to-complex-expression/test.c]: Argh, I swear I had a version of this that fixed this. The reason this is here is to trick function pointer removal (the old one) basically.

File: regression/goto-instrument/restrict-function-pointer-to-complex-expression/test.c (hunk 31 above)
Comment (hannes-steffenhagen-diffblue) [reply regression/goto-instrument/restrict-function-pointer-to-complex-expression/test.c:15]: The FP removal here happens before symex, so if constant propagation works for this we did our job right.

File: regression/goto-instrument/restrict-function-pointer-to-multiple-functions-code-check/test.desc (hunk 33 above)
Comment (hannes-steffenhagen-diffblue) [reply regression/goto-instrument/restrict-function-pointer-to-multiple-functions-code-check/test.desc:4]: What does the green heart mean?

File: regression/goto-instrument/restrict-function-pointer-to-multiple-functions-via-file-and-command-line-options/test.desc (hunk 34 above)
Comment (hannes-steffenhagen-diffblue) [reply regression/goto-instrument/restrict-function-pointer-to-multiple-functions-via-file-and-command-line-options/test.desc:3]: This is testing the error message we get from this pass, if the command line arguments were ignored you’d get a different error message or none at all.

File: src/goto-programs/label_function_pointer_call_sites.cpp (hunk 5 above)
Comment (hannes-steffenhagen-diffblue) [reply src/goto-programs/label_function_pointer_call_sites.cpp:13]: This documentation is in the header.

File: src/goto-programs/label_function_pointer_call_sites.cpp (hunk 2 above)
Comment (hannes-steffenhagen-diffblue) [reply src/goto-programs/label_function_pointer_call_sites.cpp:59]: Not in general, but goto function bodies are `std::list` which is safe to modify while iterating. We have more code that assumes this. May be worth putting a comment here that says as much though.

File: src/goto-programs/restrict_function_pointers.cpp (hunk 13 above)
Comment (hannes-steffenhagen-diffblue) [reply src/goto-programs/restrict_function_pointers.cpp:21]: This is just to hide implementation-specific symbols, same as writing `static` in front of them. I just hide functions that don’t need to be global from the global namespace out of habit.

File: src/goto-programs/restrict_function_pointers.cpp (hunk 25 above)
Comment (hannes-steffenhagen-diffblue) [reply src/goto-programs/restrict_function_pointers.cpp]: I didn’t know we had that

File: src/goto-programs/restrict_function_pointers.cpp (hunk 24 above)
Comment (hannes-steffenhagen-diffblue) [reply src/goto-programs/restrict_function_pointers.cpp:108]: I didn’t even know we had that in cbmc!

File: src/goto-programs/restrict_function_pointers.h (hunk 8 above)
Comment (hannes-steffenhagen-diffblue) [reply src/goto-programs/restrict_function_pointers.h:41]: I believe this is explained in the cprover-manual

**Review by hannes-steffenhagen-diffblue — COMMENTED** [review]
Addressing some of Thomas’s comments.

**Comment by smowton** [comment]
Looks like there's a ton of comments to look at-- let me know when this / its successor PR are in final-ish state and waiting on code owners?

File: src/goto-programs/label_function_pointer_call_sites.h (hunk 7 above)
Comment (danpoe) [reply src/goto-programs/label_function_pointer_call_sites.h:28]: Put it into a separate commit and added unit tests as well.

File: src/goto-programs/restrict_function_pointers.cpp (hunk 9 above)
Comment (danpoe) [reply src/goto-programs/restrict_function_pointers.cpp]: We now have functions `to_json()` and `from_json()`, and `write_to_file()` and `read_from_file()` use those. I've also added unit tests for the former two.

File: src/goto-programs/restrict_function_pointers.cpp (hunk 10 above)
Comment (danpoe) [reply src/goto-programs/restrict_function_pointers.cpp]: Removed this lambda.

File: src/goto-programs/restrict_function_pointers.cpp (hunk 12 above)
Comment (danpoe) [reply src/goto-programs/restrict_function_pointers.cpp:357]: Will defer this to the next PR which changes `from_options()`.

File: src/goto-programs/restrict_function_pointers.cpp (hunk 15 above)
Comment (danpoe) [reply src/goto-programs/restrict_function_pointers.cpp:268]: Done

File: src/goto-programs/restrict_function_pointers.cpp (hunk 16 above)
Comment (danpoe) [reply src/goto-programs/restrict_function_pointers.cpp]: Added

File: src/goto-programs/restrict_function_pointers.cpp (hunk 17 above)
Comment (danpoe) [reply src/goto-programs/restrict_function_pointers.cpp]: Done

File: src/goto-programs/restrict_function_pointers.cpp (hunk 24 above)
Comment (danpoe) [reply src/goto-programs/restrict_function_pointers.cpp:108]: I would prefer those functions actually. Kept them.

File: src/goto-programs/restrict_function_pointers.cpp (hunk 27 above)
Comment (danpoe) [reply src/goto-programs/restrict_function_pointers.cpp:1]: 450 lines is still within reasonable bounds IMO

File: regression/goto-instrument/restrict-function-pointer-to-complex-expression/test.desc (hunk 28 above)
Comment (danpoe) [reply regression/goto-instrument/restrict-function-pointer-to-complex-expression/test.desc:7]: Done

File: regression/goto-instrument/restrict-function-pointer-to-complex-expression/test.desc (hunk 30 above)
Comment (danpoe) [reply regression/goto-instrument/restrict-function-pointer-to-complex-expression/test.desc:3]: Will defer 1 to the next PR.

File: src/goto-programs/restrict_function_pointers.cpp (hunk 14 above)
Comment (danpoe) [reply src/goto-programs/restrict_function_pointers.cpp]: It could be, though it probably wouldn't improve things much. The current version has the advantage that specific error messages are printed for different format errors.

File: src/goto-programs/restrict_function_pointers.cpp (hunk 18 above)
Comment (danpoe) [reply src/goto-programs/restrict_function_pointers.cpp]: Moved the body of the loop into a separate function

File: src/goto-programs/restrict_function_pointers.cpp (hunk 19 above)
Comment (danpoe) [reply src/goto-programs/restrict_function_pointers.cpp]: Inverted the ifs :green_heart: and put the body of the loop into a separate function

File: doc/cprover-manual/restrict-function-pointer.md (hunk 29 above)
Comment (danpoe) [reply doc/cprover-manual/restrict-function-pointer.md:15]: Will leave refactoring the documentation to the next PR as it changes some things.

File: src/goto-programs/restrict_function_pointers.cpp (hunk 23 above)
Comment (danpoe) [reply src/goto-programs/restrict_function_pointers.cpp]: Put this into a separate function. I think it is not necessary to add unit tests as this functionality is very well tested by the regression tests already.

File: src/goto-programs/label_function_pointer_call_sites.cpp (hunk 3 above)
Comment (danpoe) [reply src/goto-programs/label_function_pointer_call_sites.cpp]: I think the original version is fine. I've added more visual breaks to make it easier to read.

File: src/goto-programs/label_function_pointer_call_sites.cpp [hunk 35]
```diff
@@ -0,0 +1,65 @@
...
+      {
+        auto const &function_call = it->get_function_call();
+        auto const &function_pointer_dereference =
+          to_dereference_expr(function_call.function());
+        auto const &source_location = function_call.source_location();
+        auto const &goto_function_symbol_mode =
+          goto_model.symbol_table.lookup_ref(goto_function.first).mode;
+
+        auto const call_site_symbol_name =
+          irep_idt{id2string(goto_function.first) + ".function_pointer_call." +
```
Comment (smowton) [inline src/goto-programs/label_function_pointer_call_sites.cpp:33]: May be best not to use `.` since it delimits members, and this isn't one ($ has been used elsewhere for a reserved character that doesn't mean anything in C)

File: src/goto-programs/restrict_function_pointers.cpp [hunk 36]
```diff
@@ -0,0 +1,478 @@
...
+{
+  typecheck_function_pointer_restrictions(goto_model, restrictions);
+
+  for(auto &function_item : goto_model.goto_functions.function_map)
+  {
+    goto_functiont &goto_function = function_item.second;
+
+    for_each_function_call(
+      goto_function,
+      std::bind(
```
Comment (smowton) [inline src/goto-programs/restrict_function_pointers.cpp]: This is one verbose way to write
```
[&](const goto_programt::targett &location) { handle_call(goto_function, goto_model, restrictions, location); }
```

**Review by smowton — APPROVED** [review]
Only skimmed the code, didn't review the docs or tests

File: src/goto-programs/label_function_pointer_call_sites.cpp (hunk 35 above)
Comment (danpoe) [reply src/goto-programs/label_function_pointer_call_sites.cpp:33]: Agreed. This will change in the next PR anyways so I'll leave this as is for now.
